packages/governance/src/votes/votes.cairo

```diff
@@ -0,0 +1,301 @@
+        /// Returns the number of voting units for a given account.
+        ///
+        /// This implementation is specific to ERC20 tokens, where the balance
+        /// of tokens directly represents the number of voting units.
+        fn get_voting_units(
+            self: @ComponentState<TContractState>, account: ContractAddress
+        ) -> u256 {
+            let mut erc20_component = get_dep_component!(self, ERC20);
+            erc20_component.balance_of(account)
+        }
+    }
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl TokenTrait: TokenVotesTrait<ComponentState<TContractState>>,
+        +NoncesComponent::HasComponent<TContractState>,
+        +SNIP12Metadata,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Delegates all of `account`'s voting units to `delegatee`.
+        ///
+        /// Emits a `DelegateChanged` event.
+        /// May emit one or two `DelegateVotesChanged` events.
+        fn _delegate(
+            ref self: ComponentState<TContractState>,
+            account: ContractAddress,
+            delegatee: ContractAddress
+        ) {
+            let from_delegate = self.delegates(account);
+            self.Votes_delegatee.write(account, delegatee);
+            self
+                .emit(
+                    DelegateChanged { delegator: account, from_delegate, to_delegate: delegatee }
+                );
+            self
+                .move_delegate_votes(
+                    from_delegate, delegatee, TokenTrait::get_voting_units(@self, account)
+                );
+        }
+
+        /// Moves delegated votes from one delegate to another.
+        ///
+        /// May emit one or two `DelegateVotesChanged` events.
+        fn move_delegate_votes(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            amount: u256
+        ) {
+            let zero_address = Zero::zero();
+            let block_timestamp = starknet::get_block_timestamp();
+            if (from != to && amount > 0) {
+                if (from != zero_address) {
+                    let mut trace = self.Votes_delegate_checkpoints.read(from);
+                    let (previous_votes, new_votes) = trace
+                        .push(block_timestamp, trace.latest() - amount);
+                    self.emit(DelegateVotesChanged { delegate: from, previous_votes, new_votes });
+                }
+                if (to != zero_address) {
+                    let mut trace = self.Votes_delegate_checkpoints.read(to);
+                    let (previous_votes, new_votes) = trace
+                        .push(block_timestamp, trace.latest() + amount);
+                    self.emit(DelegateVotesChanged { delegate: to, previous_votes, new_votes });
+                }
+            }
```

Comment:
Gustavo, there are 2 things I'd like to highlight, although they are very minor ones:
1. We don't have to wrap an `if ...` condition statement in brackets
2. There's no reason to initialize a local variable for zero address to compare other addresses to it, we can just use `some_address.is_zero()` or `some_address.is_non_zero()`

```suggestion
            let block_timestamp = starknet::get_block_timestamp();
            if from != to && amount > 0 {
                if from.is_non_zero() {
                    let mut trace = self.Votes_delegate_checkpoints.read(from);
                    let (previous_votes, new_votes) = trace
                        .push(block_timestamp, trace.latest() - amount);
                    self.emit(DelegateVotesChanged { delegate: from, previous_votes, new_votes });
                }
                if to.is_non_zero() {
                    let mut trace = self.Votes_delegate_checkpoints.read(to);
                    let (previous_votes, new_votes) = trace
                        .push(block_timestamp, trace.latest() + amount);
                    self.emit(DelegateVotesChanged { delegate: to, previous_votes, new_votes });
                }
            }
```

Reply:
Thanks! Updated, I think you meant parenthesis instead of brackets.